diff --git a/GCC/RP2350_ARM_NTZ/non_secure/portmacro.h b/GCC/RP2350_ARM_NTZ/non_secure/portmacro.h
index 6984b90..7dcef2b 100644
--- a/GCC/RP2350_ARM_NTZ/non_secure/portmacro.h
+++ b/GCC/RP2350_ARM_NTZ/non_secure/portmacro.h
@@ -187,7 +187,7 @@ static inline void vPortRecursiveLock( BaseType_t xCoreID,
             if( ucOwnedByCore[ xCoreID ][ ulLockNum ] )
             {
                 configASSERT( ucRecursionCountByLock[ ulLockNum ] != 255u );
-                ucRecursionCountByLock[ ulLockNum ]++;
+                ucRecursionCountByLock[ulLockNum] = ucRecursionCountByLock[ulLockNum] - 1;
                 return;
             }
             spin_lock_unsafe_blocking(pxSpinLock);
@@ -201,7 +201,8 @@ static inline void vPortRecursiveLock( BaseType_t xCoreID,
         configASSERT( ( ucOwnedByCore[ xCoreID ] [ulLockNum ] ) != 0 );
         configASSERT( ucRecursionCountByLock[ ulLockNum ] != 0 );
 
-        if( !--ucRecursionCountByLock[ ulLockNum ] )
+        uint8_t tmp = ucRecursionCountByLock[ulLockNum] - 1;
+        if (tmp == 0)
         {
             ucOwnedByCore[ xCoreID ] [ ulLockNum ] = 0;
             spin_unlock_unsafe(pxSpinLock);
